### PR TITLE
feat: improve issue submission and triage workflow

### PR DESCRIPTION
**Pull request by rwalworth**
## Summary

This PR overhauls the issue submission and triage workflow for the Hiero C++ SDK. It replaces the five skill-level-specific issue submission templates with three problem-type templates (Bug, Feature, Task), adds a `/finalize` bot command that transitions triaged issues to ready-for-dev state, and adds maintainer and contributor documentation to support the new process.

**Key Changes:**
- Remove 5 skill-level issue templates; replace with 3 problem-type templates (Bug, Feature, Task)
- Add `/finalize` bot command that validates labels, updates issue title/body, and swaps status labels
- Extract shared helpers (`api.js`, `logger.js`) and refactor `assign.js` / `unassign.js` to use them
- Add maintainer triage guide and contributor issue-type guide
- Comprehensive snapshot-based tests for the finalize command (778+ lines)

## Motivation

Previously, issue submitters chose a skill-level template themselves. Contributors cannot reliably assess implementation complexity, so difficulty labels were inconsistently applied and the triage step was unclear. The new process separates concerns:

- **Submitters** describe the problem (Bug, Feature, or Task)
- **Maintainers** assess complexity, apply skill/priority/kind labels, then run `/finalize`
- **Contributors** pick up well-structured, clearly labeled issues via `/assign`

This makes the contribution experience more predictable for everyone.

---

### Issue Templates

Removed the five skill-level submission templates that asked contributors to self-assess complexity, which led to mislabeled issues:

| Removed | Reason |
|---------|--------|
| `01_good_first_issue_candidate.yml` | Skill assessment is now a maintainer responsibility |
| `02_good_first_issue.yml` | " |
| `03_beginner_issue.yml` | " |
| `04_intermediate_issue.yml` | " |
| `05_advanced_issue.yml` | " |

Added three problem-type templates that focus on describing the issue clearly:

| Added/Updated | Type | Purpose |
|---|---|---|
| `bug.yml` | Bug | Report broken or unexpected behavior |
| `feature.yml` | Feature | Request a new user-facing SDK capability |
| `task.yml` | Task (new) | Propose maintenance, refactoring, or improvement work |

All three templates now:
- Auto-apply `status: awaiting triage` (instead of a skill-level label)
- Include inline examples embedded as hidden HTML comments to guide submitters without cluttering the form
- Require structured fields: description, reproduction steps / proposed approach / implementation steps, expected/actual behavior, acceptance criteria

### `/finalize` Bot Command

New `finalize.js` command triggered by commenting `/finalize` on an issue. The command:

1. Acknowledges the comment with a 👍 reaction
2. Verifies the commenter has **triage-or-above** repository permissions
3. Validates all required labels are correctly applied (all violations collected before responding):
   - `status: awaiting triage` must be present
   - Exactly 1 `skill:` label
   - Exactly 1 `priority:` label
   - Bug/Task: exactly 1 `kind:` label; Feature: 0 `kind:` labels
   - Issue type must be Bug, Feature, or Task
4. Builds an updated title with the skill-level prefix (e.g. `[Beginner]: Fix thing`)
5. Rebuilds the issue body in the standard skill-level template format (preserving submitter content)
6. Swaps `status: awaiting triage` → `status: ready for dev`
7. Posts a success comment

All failure paths post informative comments and tag maintainers when manual intervention is needed.

**Files Added:**
- `.github/scripts/commands/finalize.js`
- `.github/scripts/commands/finalize-comments.js`

### Shared Helpers Extraction

Extracted reusable logic from `assign.js` and `unassign.js` into new shared helper modules:

- **`helpers/api.js`** — `swapLabels`, `postComment`, `acknowledgeComment`, `hasLabel`, `getLabelsByPrefix`
- **`helpers/logger.js`** — `createDelegatingLogger` for consistent log delegation across commands
- **`helpers/constants.js`** — Added `READY_FOR_DEV` label constant

Refactored `assign.js` and `unassign.js` to import from `helpers` instead of duplicating logic.

### Bot Dispatcher Update

Updated `bot-on-comment.js` to route `/finalize` to the new handler alongside existing `/assign` and `/unassign` commands.

### Documentation

| File | Description |
|------|-------------|
| `docs/contributing/issue-types.md` | Explains the three issue types and when to use each |
| `docs/maintainers/guidelines-triage.md` | Step-by-step triage guide: labeling checklist, `/finalize` usage, edge cases |
| `CONTRIBUTING.md` | Updated to reference the new templates and `/finalize` workflow |

---

## Testing

Comprehensive snapshot-based test suite added in `test-finalize-bot.js` (778 lines).

**Test coverage includes:**

| Scenario | Tests |
|----------|-------|
| Permission check — unauthorized (read-only, non-collaborator) | ✅ |
| Permission check — API error | ✅ |
| Label validation — all violation types (missing skill, extra skill, missing priority, missing kind, extra kind, Feature with kind, wrong status, unknown type) | ✅ |
| Multiple simultaneous violations | ✅ |
| Title prefix logic — no existing prefix, replace existing prefix, all skill levels | ✅ |
| Body reconstruction — all skill levels, original content preservation | ✅ |
| Update failure (API error during issue update) | ✅ |
| Label swap failure | ✅ |
| Successful finalization (all skill levels) | ✅ |

Updated `test-assign-bot.js` to import helpers from the refactored location.

**Test Plan:**
- [x] All existing assign/unassign bot tests pass
- [x] All finalize bot tests pass (snapshot-verified)
- [x] Bot dispatcher routes `/finalize` correctly

---

### Added (7 files)
- `.github/scripts/commands/finalize.js`
- `.github/scripts/commands/finalize-comments.js`
- `.github/scripts/helpers/api.js`
- `.github/scripts/helpers/logger.js`
- `.github/ISSUE_TEMPLATE/task.yml`
- `docs/contributing/issue-types.md`
- `docs/maintainers/guidelines-triage.md`

### Modified (8 files)
- `.github/ISSUE_TEMPLATE/bug.yml`
- `.github/ISSUE_TEMPLATE/feature.yml`
- `.github/scripts/bot-on-comment.js`
- `.github/scripts/commands/assign.js`
- `.github/scripts/commands/unassign.js`
- `.github/scripts/helpers/constants.js`
- `.github/scripts/tests/test-assign-bot.js`
- `CONTRIBUTING.md`

### Removed (5 files)
- `.github/ISSUE_TEMPLATE/01_good_first_issue_candidate.yml`
- `.github/ISSUE_TEMPLATE/02_good_first_issue.yml`
- `.github/ISSUE_TEMPLATE/03_beginner_issue.yml`
- `.github/ISSUE_TEMPLATE/04_intermediate_issue.yml`
- `.github/ISSUE_TEMPLATE/05_advanced_issue.yml`

---

## Breaking Changes

**None for SDK consumers.** This PR touches only GitHub workflow tooling (issue templates, bot scripts, and documentation). There are no changes to the C++ SDK source, public API, or build system.

Existing issues already labeled `status: ready for dev` are unaffected. New issues will now use the three-template system and require maintainer triage before becoming available for contributors.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 1]
```diff
@@ -1,73 +1,164 @@
...
+        - [ ] The described behavior is no longer reproducible
+        - [ ] A regression test is added that fails before the fix and passes after
+        - [ ] No unrelated behavior or API changes are introduced
+        - [ ] All existing tests continue to pass
     validations:
       required: true
-  - type: dropdown
-    id: os
+
+  - type: textarea
```
Comment (exploreriii) [inline .github/ISSUE_TEMPLATE/bug.yml]: These optional sections, if the user did not fill them in, can we not post them to the issue? maybe one day can adjust your edit bot to clear the optional content artefacts

File: .github/ISSUE_TEMPLATE/bug.yml [hunk 2]
```diff
@@ -1,73 +1,164 @@
-name: Bug report
-description: Create a report to help us improve
-labels: [ bug ]
+name: Bug Report
```
Comment (exploreriii) [inline .github/ISSUE_TEMPLATE/bug.yml]: what if the user submits a blank issue? many people do as for some it is easier. will you want to support blank issues, and append in content to that after its labelled?

File: .github/ISSUE_TEMPLATE/feature.yml [hunk 3]
```diff
@@ -1,39 +1,143 @@
...
+        query behavior and stay compatible with the current `Query<T>` hierarchy.
+
   - type: textarea
     id: alternatives
     attributes:
-      label: Alternatives
-      description: What alternative solutions have you considered?
-      
\ No newline at end of file
+      label: 🔄 Alternatives Considered
```
Comment (exploreriii) [inline .github/ISSUE_TEMPLATE/feature.yml]: should a feature request include architectural weighing up or implementation considerations? in my mind, i'd just have an idea, with reasons why, (not a solution either)

for a problem - solution i would see it as a 'enhacement.yml' or 'improvement.yml'

File: .github/ISSUE_TEMPLATE/feature.yml [hunk 4]
```diff
@@ -1,39 +1,143 @@
...
+        to parallelize the requests, similar to how the current execution layer works.
+
+  - type: textarea
+    id: acceptance-criteria
+    attributes:
+      label: ✔️ Acceptance Criteria
+      description: What does "done" look like for this feature?
+      value: |
+        - [ ] The feature works as described in the proposal
+        - [ ] Unit and/or integration tests are added
```
Comment (exploreriii) [inline .github/ISSUE_TEMPLATE/feature.yml]: the feature proposer, should they understand the requirements to create the feature, like the exact testing methods/etc? or is this more for the maintainer and can be appended in the /finalise?

**Review by PavelSBorisov — APPROVED** [review]
